### PR TITLE
Implement CP Subsystem persistence

### PR DESCRIPTION
**Pull request by mdogan**
See individual commits for details.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3192

### CODE REVIEW

**Comment by mmedenjak** [comment]
Great work guys, congrats!

File: hazelcast/src/main/java/com/hazelcast/config/cp/RaftAlgorithmConfig.java [hunk 1]
```diff
@@ -63,47 +63,79 @@
...
+     * taken by the local Raft node. This value must be configured wisely as it
+     * effects performance of the system in multiple ways. If a small value is
+     * set, it means that snapshots are taken too frequently and Raft nodes keep
+     * a very short Raft log. If snapshots are large and CP Subsystem
+     * Persistence is enabled, this can create an unnecessary overhead on IO
+     * performance. Moreover, a Raft leader can send too many snapshots to
+     * followers and this can create an unnecessary overhead on network.
+     * On the other hand, if a very large value is set, it can create a memory
+     * overhead since Raft log entries are going to be kept in memory until
+     * the next snapshot.
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/config/cp/RaftAlgorithmConfig.java]: This is actually in Raft sense, independent of any operations, it's about advance of Raft `commitIndex`.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java [hunk 2]
```diff
@@ -968,6 +1203,96 @@ public void toLeader() {
...
+            runQuery(t.element1, t.element2);
+        }
+
+        queryState.reset();
+        return false;
+    }
+
     private void appendEntryAfterLeaderElection() {
         Object entry = raftIntegration.getAppendedEntryOnLeaderElection();
         if (entry != null) {
```
[the remark this answers is not in the document]
Comment (metanet) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java]: We can make a flush here, but we don't really need it. In our Raft code, Raft leaders do not flush entries eagerly. An entry is flushed on the leader after it is sent to a follower. We can even commit an entry even before it is flushed on the leader.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftEndpointImpl.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.UUID;
+
+import static com.hazelcast.internal.util.UUIDSerializationUtil.readUUID;
+import static com.hazelcast.internal.util.UUIDSerializationUtil.writeUUID;
+
+/**
+ * Represents an endpoint that runs the Raft consensus algorithm as a member of
+ * a Raft group.
+ */
+public class RaftEndpointImpl implements RaftEndpoint, IdentifiedDataSerializable, Serializable {
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftEndpointImpl.java]: We discussed the same. Advantage of using `RaftEndpoint[Impl]` is, it's open to upgrades. Later we can add new fields, attributes to the endpoint class. But if it's a `UUID`, then we may need more complex changes in future. That's why we finally opted to use `RaftEndpoint`.

File: hazelcast/src/main/java/com/hazelcast/cp/CPSubsystem.java [hunk 4]
```diff
@@ -22,135 +22,352 @@
...
+ * Subsystem Persistence is disabled, hence a shut-down CP member cannot join
+ * back with its CP identity and state, hence it is better to remove it from CP
+ * Subsystem to not to harm availability of CP groups. If there is no other
+ * available CP member to replace a shutting down CP member in a CP group, that
+ * CP group's size is reduced by 1 and its majority value is recalculated.
+ * On the other hand, when CP Subsystem Persistence is enabled, a shut-down CP
+ * member can come back by restoring its CP state. Therefore, it is not
+ * automatically removed from CP Subsystem when CP Subsystem Persistence is
+ * enabled. It is up to the user to remove shut-down CP members
+ * via {@link CPSubsystemManagementService#removeCPMember(String)} if they will
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/cp/CPSubsystem.java]: will fix 👍

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java [hunk 5]
```diff
@@ -267,26 +312,48 @@ public void restartLocal(long seed) {
...
-        resetLocalRaftState();
+        nodeLock.writeLock().lock();
+        try {
+            // we should clear the current raft state before resetting the metadata manager
+            resetLocalRaftState();
 
-        metadataGroupManager.restart(seed);
-        logger.info("CP state is reset with groupId seed: " + seed);
+            getCPPersistenceService().reset();
+            metadataGroupManager.restart(seed);
```
[the remark this answers is not in the document]
Comment (metanet) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java]: Good catch. Unfortunately, CP Subsystem reset is not a 100% safe process. We state this in its Javadoc. It is successful only if it is done on all members of the cluster without any failure. If any failure occurs, the user needs to do a manual check. Maybe we can also state in the docs that persistence files must be deleted manually on a failed node before its restart...

File: hazelcast/src/main/java/com/hazelcast/config/cp/RaftAlgorithmConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/config/cp/RaftAlgorithmConfig.java]: Got it, thanks.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java]: Got it. Thanks.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftEndpointImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftEndpointImpl.java]: I see, thanks.

File: hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java]: But is it be possible to distinguish between valid and invalid states (or at least some of them) by looking at the list of files present? If we expect that the user should delete files, how and when do we inform him that it is needed?

**Comment by devozerov** [comment]
As discussed, concerns about recovery will be addressed in separate PRs during 4.0 stabilization phase.

**Comment by mdogan** [comment]
run-lab-run

**Pull request merged**